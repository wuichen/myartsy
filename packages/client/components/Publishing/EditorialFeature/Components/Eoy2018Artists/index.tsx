import { Box, color, Flex, media, Sans, Serif, space } from "@artsy/palette"
import { compact, find, map } from "lodash"
import React from "react"
import styled from "styled-components"

import { unica } from "Assets/Fonts"
import { Byline, BylineContainer } from "Components/Publishing/Byline/Byline"
import { EditorialFeaturesProps } from "Components/Publishing/EditorialFeature/EditorialFeature"
import { Nav, NavContainer } from "Components/Publishing/Nav/Nav"
import { ArticleCardContainer } from "Components/Publishing/RelatedArticles/ArticleCards/ArticleCard"
import {
  ArticleCards,
  ArticlesWrapper,
} from "Components/Publishing/RelatedArticles/ArticleCards/ArticleCards"
import {
  FullLabel,
  ImageSetContainer,
  ImageSetPreview,
  ImgContainer,
} from "Components/Publishing/Sections/ImageSetPreview"
import {
  LabelWrapper,
  SlideshowCta,
  SlideshowTitle,
} from "Components/Publishing/Sections/ImageSetPreview/ImageSetLabel"
import { StyledText } from "Components/Publishing/Sections/StyledText"
import { Text } from "Components/Publishing/Sections/Text"
import { resize } from "Utils/resizer"
import { Eoy2018ArticleHeader } from "./ArticleHeader"

export class Eoy2018Artists extends React.Component<EditorialFeaturesProps> {
  getHeaderSections = () => {
    const { sections } = this.props.article
    const headers = []

    sections.map((section, index) => {
      let headerSection
      const sectionAfter = sections[index + 1]
      const sectionAfterisImage =
        sectionAfter && sectionAfter.type === "image_collection"

      if (this.sectionIsHeader(section)) {
        headerSection = {
          index,
          section,
        }
        if (sectionAfterisImage) {
          headerSection.imageSection = sectionAfter
        }
        headers.push(headerSection)
      }
    })
    return headers
  }

  sectionIsHeader = section => {
    const isText = section.type === "text"
    const isHeader = isText && section.body.includes("<h1>")
    return isHeader
  }

  sectionArtistHeader = (section, i) => {
    const headerSections = this.getHeaderSections()
    const { imageSection } = find(headerSections, ["section", section])
    const src = imageSection && imageSection.images[0].url
    const caption = imageSection && imageSection.images[0].caption

    return (
      <ArtistHeaderSection
        key={i}
        mb={40}
        flexDirection={["column", "column", "row", "row"]}
      >
        <ArtistHeaderTitle dangerouslySetInnerHTML={{ __html: section.body }} />
        <ArtistHeaderImg
          src={src && resize(src, { width: 700 })}
          alignItems="flex-end"
        >
          {caption && (
            <Sans size="1" color="white" px={10} py={5}>
              <div dangerouslySetInnerHTML={{ __html: caption }} />
            </Sans>
          )}
        </ArtistHeaderImg>
      </ArtistHeaderSection>
    )
  }

  sectionText = (section, i) => {
    const {
      article: { sections, layout },
      showTooltips,
    } = this.props
    const isChapterStart =
      sections[i - 1] && sections[i - 1].type === "image_collection"

    return (
      <Box maxWidth={["100%", "75%", "75%"]} ml="auto" px={[20, 0]} key={i}>
        <TextSection size="5" pb={[40, 60]} isChapterStart={isChapterStart}>
          <Text
            html={section.body}
            layout={layout}
            showTooltips={showTooltips}
          />
        </TextSection>
      </Box>
    )
  }

  sectionImageSet = (section, i) => {
    return (
      <ImageSetWrapper
        key={i}
        mb={60}
        maxWidth={["100%", "100%", "100%", "75%"]}
      >
        <ImageSetPreview section={section}>
          <CaptionWrapper size={["3", "4"]}>
            <ImageSetCaption
              dangerouslySetInnerHTML={{ __html: section.images[0].caption }}
            />
          </CaptionWrapper>
        </ImageSetPreview>
      </ImageSetWrapper>
    )
  }

  getSections = () => {
    const {
      article: { sections },
    } = this.props
    const headerSections = this.getHeaderSections()
    const headerTextIndexes = map(headerSections, "index")
    const headerImages = map(headerSections, "image")

    const renderedSections = sections.map((section, i) => {
      const isHeader = headerTextIndexes.includes(i)
      const isHeaderImage = headerImages.includes(section)

      if (isHeader) {
        return this.sectionArtistHeader(section, i)
      } else if (isHeaderImage) {
        return null
      } else {
        if (section.type === "text" && i !== 0) {
          return this.sectionText(section, i)
        } else if (section.type === "image_set") {
          return this.sectionImageSet(section, i)
        } else {
          return null
        }
      }
    })
    return renderedSections
  }

  render() {
    const { article, isMobile, isTablet, isTest } = this.props
    const introText = this.sectionText(article.sections[0], 0)
    const headerImages = map(
      compact(map(this.getHeaderSections(), "imageSection")),
      "images"
    )

    return (
      <AricleWrapper>
        <Nav canFix color={color("black100")} backgroundColor="white">
          <NavBorder />
        </Nav>

        <Box px={[10, 10, 55]} maxWidth={1600} mx="auto">
          <Eoy2018ArticleHeader
            images={headerImages}
            isMobile={isMobile}
            isTablet={isTablet}
            isTest={isTest}
          />

          <ArticleContent py={40}>
            <IntroSection
              alignItems="flex-start"
              flexDirection={["column", "row"]}
              pl={[0, 20]}
            >
              <Byline article={article} />
              {introText}
            </IntroSection>
            {this.getSections()}
          </ArticleContent>
        </Box>

        {article.relatedArticles && (
          <Box px={[10, 10, 55]} mb={40} mx="auto">
            <RelatedArticleWrapper>
              <ArticleCards relatedArticles={article.relatedArticles} />
            </RelatedArticleWrapper>
          </Box>
        )}
      </AricleWrapper>
    )
  }
}

const BORDER_WIDTH = 6

const AricleWrapper = styled.div`
  ${NavContainer} {
    border-bottom: none;
  }
`

const NavBorder = styled.div`
  border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};
  position: absolute;
  top: 100%;
  left: 55px;
  right: 55px;
  max-width: calc(1600px - 110px);
  margin: 0 auto;

  ${media.md`
    left: 10px;
    right: 10px;
  `};
`

const ArticleContent = styled(Box)`
  border-left: ${BORDER_WIDTH}px solid ${color("purple100")};
  border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};

  blockquote {
    ${unica("s34")};
    line-height: 1.3em;
    margin: 0;
    font-weight: inherit;

    ${media.sm`
      ${unica("s25")};
      line-height: 1.3em;
    `};
  }
`

export const ArtistHeaderSection = styled(Flex)`
  height: 60vh;
  min-height: 450px;
  border: ${BORDER_WIDTH}px solid ${color("purple100")};
  border-left-width: 0;

  ${media.md`
    height: fit-content;
  `};
`

const ArtistHeaderTitle = styled.div`
  flex: 1;
  border-right: ${BORDER_WIDTH}px solid ${color("purple100")};
  overflow: hidden;
  min-height: min-content;

  h1 {
    ${unica("s65")};
    width: 100%;
    height: 50%;
    min-height: fit-content;
    padding: ${space(2)}px;
    border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};
    margin: 0;
    font-weight: inherit;

    ${media.md`
      width: 60%;
      float: right;
      height: 100%;
      border-bottom: none;
      border-left: ${BORDER_WIDTH}px solid ${color("purple100")};
    `};

    ${media.xs`
      ${unica("s45")};
      border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};
      border-left: none;
      width: 100%;
      height: 50%;
      float: none;
    `};
  }

  h2 {
    ${unica("s25")};
    width: 50%;
    height: 50%;
    min-height: fit-content;
    display: inline-flex;
    padding: ${space(2)}px;
    margin: 0;
    font-weight: inherit;

    &:last-child {
      border-left: ${BORDER_WIDTH}px solid ${color("purple100")};
    }

    ${media.md`
      width: 40%;
      float: left;
      &:last-child {
        border-left: none;
        border-top: ${BORDER_WIDTH}px solid ${color("purple100")};
      }
    `};

    ${media.xs`
      ${unica("s19")};
      width: 50%;
      float: none;
      &:last-child {
        border-left: ${BORDER_WIDTH}px solid ${color("purple100")};
        border-top: none;
      }
    `};
  }

  ${media.md`
    height: 65vh;
    border-right: none;
    border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};
  `};
`

const ArtistHeaderImg = styled(Flex)<{ src?: string }>`
  flex: 1;
  background: ${color("purple100")};

  p {
    text-shadow: 0 0 5px black;
    opacity: 0.6;
  }

  ${props =>
    props.src &&
    `
    background: url(${props.src});
    background-size: cover;
    background-position: center;
  `};

  ${media.md`
    min-height: 70vw;
  `};
`

const IntroSection = styled(Flex)`
  ${BylineContainer} {
    flex-direction: column;
    align-items: flex-start;
    ${media.md`
      padding: 0 20px 20px;
    `};
  }
`

const TextSection = styled(Serif)<{ isChapterStart?: boolean }>`
  ${StyledText} {
    blockquote {
      padding: 0;
    }

    p,
    .paragraph {
      font-size: 24px;
      text-indent: 2em;
      padding: 0;

      ${props =>
        props.isChapterStart &&
        `
        &:first-child {
          text-indent: 0;
        }
    `};
    }
  }
`

const CaptionWrapper = styled(Sans)`
  flex: 1;

  ${media.sm`
    min-width: 50%;
  `};
`

const ImageSetCaption = styled.div`
  height: 100%;
  padding: ${space(2)}px;
  background: white;
  color: black;
  border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};

  ${media.sm`
    border-bottom: none;
    border-right: ${BORDER_WIDTH}px solid ${color("purple100")};
  `};
`

const ImageSetWrapper = styled(Box)`
  border: ${BORDER_WIDTH}px solid ${color("purple100")};
  border-left: none;

  ${ImageSetContainer} {
    display: flex;
    flex-direction: row-reverse;

    ${media.sm`
      flex-direction: column-reverse;
    `};
  }

  ${FullLabel} {
    position: relative;
    display: flex;
    flex-direction: column-reverse;
    flex: 1;
    left: auto;
    bottom: auto;
    max-width: inherit;
    border-radius: 0;
    box-shadow: none;

    &:hover {
      background-color: ${color("purple100")};
    }
    ${media.sm`
      flex-direction: row-reverse;

      ${LabelWrapper} {
        max-width: 50%;
      }
    `};
  }

  ${SlideshowTitle} {
    display: none;
  }

  ${SlideshowCta} {
    flex-direction: column;

    > div {
      padding: 0;
    }
  }

  ${ImgContainer} {
    flex: 3;
    border-right: ${BORDER_WIDTH}px solid ${color("purple100")};

    img {
      object-fit: cover;
      object-position: center;
      height: 100%;
    }

    ${media.sm`
      border-right: none;
      border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};
    `};
  }
`

const RelatedArticleWrapper = styled.div`
  border-right: ${BORDER_WIDTH}px solid ${color("purple100")};
  border-bottom: ${BORDER_WIDTH}px solid ${color("purple100")};

  ${ArticlesWrapper} {
    margin-bottom: 0;
    margin-top: 0;
  }

  ${ArticleCardContainer} {
    border: none;
  }
`
