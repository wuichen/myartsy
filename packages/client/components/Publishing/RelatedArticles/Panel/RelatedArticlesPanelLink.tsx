import { color, Serif, space } from "@artsy/palette"
import { track } from "Artsy/Analytics"
import * as Schema from "Artsy/Analytics/Schema"
import { RelatedArticlePanelData } from "Components/Publishing/Typings"
import React from "react"
import styled from "styled-components"
import { crop } from "../../../../Utils/resizer"
import { getEditorialHref } from "../../Constants"

interface RelatedArticlesPanelProps extends React.HTMLProps<HTMLDivElement> {
  article: RelatedArticlePanelData
}

@track()
export class RelatedArticlesPanelLink extends React.Component<
  RelatedArticlesPanelProps
> {
  static defaultProps = {
    label: "Related Stories",
  }

  @track<RelatedArticlesPanelProps>(props => ({
    action_type: Schema.ActionType.Click,
    destination_path: getEditorialHref(
      props.article.layout,
      props.article.slug
    ),
    // TODO: add type to schema
    type: "thumbnail",
  }))
  onClick(e) {
    // noop
  }

  render() {
    const { article } = this.props
    const href = getEditorialHref(article.layout, article.slug)
    const articleImageSrc = crop(article.thumbnail_image, {
      width: 160,
      height: 110,
    })

    return (
      <ArticleLink href={href} onClick={this.onClick.bind(this)}>
        <ArticleImage src={articleImageSrc} />
        <Serif size="4t" color={color("black100")}>
          {article.thumbnail_title}
        </Serif>
      </ArticleLink>
    )
  }
}

export const ArticleLink = styled.a`
  text-decoration: none;
  display: flex;
  justify-content: left;
  margin-bottom: ${space(2)}px;
`

const ArticleImage = styled.img`
  min-width: 80px;
  height: 55px;
  margin-right: ${space(1)}px;
`
