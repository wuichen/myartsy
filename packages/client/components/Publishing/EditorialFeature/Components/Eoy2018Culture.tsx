import { Box, color, Flex, Sans, Serif } from "@artsy/palette"
import { unica } from "Assets/Fonts"
import { media, pMedia } from "Components/Helpers"
import { Author, StyledAuthor } from "Components/Publishing/Byline/Author"
import { Date, DateContainer } from "Components/Publishing/Byline/Date"
import { Share } from "Components/Publishing/Byline/Share"
import { getArticleFullHref } from "Components/Publishing/Constants"
import { EditorialFeaturesProps } from "Components/Publishing/EditorialFeature/EditorialFeature"
import { Nav, NavContainer } from "Components/Publishing/Nav/Nav"
import { ArticleCards } from "Components/Publishing/RelatedArticles/ArticleCards/ArticleCards"
import { CaptionContainer } from "Components/Publishing/Sections/Caption"
import {
  ImageCollection,
  ImageCollectionItem,
} from "Components/Publishing/Sections/ImageCollection"
import {
  ImageSetContainer,
  ImageSetPreview,
} from "Components/Publishing/Sections/ImageSetPreview"
import {
  SlideshowCta,
  SlideshowTitle,
} from "Components/Publishing/Sections/ImageSetPreview/ImageSetLabel"
import { SocialEmbed } from "Components/Publishing/Sections/SocialEmbed"
import { StyledText } from "Components/Publishing/Sections/StyledText"
import { Text } from "Components/Publishing/Sections/Text"
import { ToolTipContainer } from "Components/Publishing/ToolTip/ToolTip"
import { SectionData } from "Components/Publishing/Typings"
import React from "react"
import Waypoint from "react-waypoint"
import styled from "styled-components"

interface State {
  stickyHeader: SectionData | null
}

export class Eoy2018Culture extends React.Component<
  EditorialFeaturesProps,
  State
> {
  state = {
    stickyHeader: null,
  }

  sectionText = (section, i, isDark) => {
    const {
      showTooltips,
      article: { layout },
    } = this.props
    return (
      <TextContainer key={i} px={[10, 10, 55]} py={30} mx="auto">
        <Box width={["100%", "100%", "70%"]} mx="auto" maxWidth={1000}>
          <Serif size="5">
            <Text
              html={section.body}
              layout={layout}
              showTooltips={showTooltips}
              color={isDark ? "white" : color("black100")}
            />
          </Serif>
        </Box>
      </TextContainer>
    )
  }

  sectionHeaderText = (section, i) => {
    return (
      <SectionHeader key={i} pt={10} pb={[10, 10, 20]} px={[10, 10, 55]}>
        <div dangerouslySetInnerHTML={{ __html: section.body }} />
      </SectionHeader>
    )
  }

  sectionSocialEmbed = (section, i) => {
    return (
      <SectionWrapper key={i} px={[10, 10, 55]} py={40}>
        <SocialEmbed section={section} />
      </SectionWrapper>
    )
  }

  sectionImageCollection = (section, i) => {
    return (
      <ImageWrapper key={i}>
        <ImageCollection
          sectionLayout={section.layout}
          articleLayout="feature"
          images={section.images}
          targetHeight={500}
          gutter={10}
        />
      </ImageWrapper>
    )
  }

  sectionImageSet = (section, i) => {
    return (
      <ImageSetWrapper key={i} px={[0, 0, 55]}>
        <ImageSetPreview section={section} />
      </ImageSetWrapper>
    )
  }

  sectionIsHeader = section => {
    const isText = section.type === "text"
    const isHeader = isText && section.body.includes("<h1>")
    return isHeader
  }

  makeSectionArrays = sections => {
    const chapters = []
    let currentChapter = null

    sections.map((section, i) => {
      if (this.sectionIsHeader(section)) {
        if (currentChapter === null) {
          currentChapter = 0
        } else {
          currentChapter += 1
        }
      }

      if (i !== 0) {
        if (chapters.length && chapters[currentChapter]) {
          chapters[currentChapter].push(section)
        } else {
          chapters[currentChapter] = [section]
        }
      }
    })
    return chapters
  }

  getSections = (sections, isDark) => {
    return sections.map((section, i) => {
      if (this.sectionIsHeader(section)) {
        return this.sectionHeaderText(section, i)
      } else {
        switch (section.type) {
          case "text": {
            return this.sectionText(section, i, isDark)
          }
          case "image_collection": {
            return this.sectionImageCollection(section, i)
          }
          case "image_set": {
            return this.sectionImageSet(section, i)
          }
          case "social_embed": {
            return this.sectionSocialEmbed(section, i)
          }
          default: {
            return null
          }
        }
      }
    })
  }

  render() {
    const { article } = this.props
    const chapters = this.makeSectionArrays(article.sections)
    const introText =
      article.sections && article.sections[0] && article.sections[0].body

    return (
      <ArticleWrapper>
        <Nav canFix color="black" backgroundColor="white">
          {this.state.stickyHeader && (
            <StickyHeader>
              {this.sectionHeaderText(this.state.stickyHeader, 0)}
            </StickyHeader>
          )}
        </Nav>
        <SectionWrapper>
          <ArticleTitle
            size="16"
            weight="medium"
            px={[10, 10, 55]}
            pb={10}
            pt={20}
          >
            <div>Year In Visual Culture</div>
            <div>2018</div>
          </ArticleTitle>
          <HeaderImg src={article.hero_section.url} />
        </SectionWrapper>

        <TextContainer px={[10, 10, 55]} py={30} mx="auto">
          <Box maxWidth={1400} mx="auto">
            <Flex
              pb={20}
              alignItems="center"
              flexDirection={["row-reverse", "row"]}
            >
              <Box width={["100%", "100%", "15%"]}>
                <Share
                  url={getArticleFullHref(article.slug)}
                  title={article.social_title || article.thumbnail_title}
                />
              </Box>
              <Flex width={["100%", "100%", "70%"]} maxWidth={1000}>
                <Author authors={article.authors} />
                <Date date={article.published_at} />
              </Flex>
            </Flex>

            <Box width={["100%", "100%", "70%"]} maxWidth={1000} mx="auto">
              {introText && (
                <div dangerouslySetInnerHTML={{ __html: introText }} />
              )}
            </Box>
          </Box>
          <Waypoint
            onEnter={({ previousPosition }) => {
              if (previousPosition === "above") {
                this.setState({ stickyHeader: null })
              }
            }}
          />
        </TextContainer>

        {chapters.map((chapter, i) => {
          const isDark = i % 2 === 0
          return (
            <ChapterWrapper isDark={isDark} key={i}>
              <Waypoint
                onLeave={({ currentPosition, previousPosition }) => {
                  if (
                    previousPosition === "inside" &&
                    currentPosition === "above"
                  ) {
                    this.setState({ stickyHeader: chapter[0] })
                  }
                }}
              />
              {this.getSections(chapter, isDark)}
              <Waypoint
                onEnter={({ previousPosition }) => {
                  if (previousPosition === "above") {
                    this.setState({ stickyHeader: chapter[0] })
                  }
                }}
              />
            </ChapterWrapper>
          )
        })}

        {article.relatedArticles && (
          <Box px={[10, 10, 55]} my={40} mx="auto">
            <ArticleCards relatedArticles={article.relatedArticles} />
          </Box>
        )}
      </ArticleWrapper>
    )
  }
}

const BORDER_WIDTH = 6

const ArticleWrapper = styled.div`
  ${NavContainer} {
    border-bottom-width: ${BORDER_WIDTH}px;
    border-top: ${BORDER_WIDTH}px solid black;
  }
  ${DateContainer} {
    margin-top: 0;
  }
  ${StyledAuthor} {
    margin-top: 0;

    ${media.xs`
      width: max-content;
    `};
  }
`

const ChapterWrapper = styled.div<{ isDark?: boolean }>`
  ${props =>
    props.isDark &&
    `
      color: white;
      background-color: ${color("black100")};
  `};

  ${ToolTipContainer} {
    color: ${color("black100")};
  }
`

const SectionWrapper = styled(Box)`
  border-bottom: ${BORDER_WIDTH}px solid;
  position: relative;
`

const TextContainer = styled(SectionWrapper)`
  blockquote {
    ${unica("s34")};
    padding: 20px 0;
    line-height: 1.25em;
    margin: 0;
    font-weight: inherit;
  }

  ${StyledText} {
    p,
    .paragraph {
      font-size: 24px;
      text-indent: 2em;
      padding: 0;

      &:first-child {
        text-indent: 0;
      }
    }
  }

  ${media.sm`
    blockquote {
      font-size: 24px;
    }
  `};
`

const ArticleTitle = styled(Sans)`
  text-transform: uppercase;
  border-bottom: ${BORDER_WIDTH}px solid;
  font-size: 7.5vw;
  line-height: 1em;
`

const SectionHeader = styled(Box)<{ isDark?: boolean }>`
  border-bottom: ${BORDER_WIDTH}px solid;

  > div {
    max-width: 1600px;
    margin: 0 auto;
    width: 100%;
    display: flex;
    align-items: flex-end;
  }

  h1 {
    ${unica("s65", "medium")};
    text-transform: uppercase;
    flex: 2;
    margin: 0;
    font-weight: inherit;
  }
  h2 {
    font-size: 50px;
    width: fit-content;
    line-height: 1em;
    text-align: right;
    margin: 0;
    font-weight: inherit;
  }

  ${media.md`
    h1 {
      font-size: 45px;;
    }
    h2 {
      font-size: 30px;
    }
  `};

  ${media.xs`
    h1 {
      font-size: 30px;;
    }
    h2 {
      font-size: 20px;
    }
`};
`

const HeaderImg = styled.img`
  height: calc(100vh - 50px);
  min-height: 400px;
  object-fit: cover;
  object-position: center;
  width: 100%;
`

const ImageWrapper = styled(SectionWrapper)`
  img {
    max-height: calc(100vh - 50px);
    min-height: 400px;
    object-fit: cover;
    object-position: center;

    ${media.md`
      min-height: unset;
    `};
  }

  ${ImageCollectionItem} {
    ${pMedia.xs`
      margin-bottom: 0;
      min-height: unset;
    `};
  }

  ${CaptionContainer} {
    position: absolute;
    bottom: 0;
    left: 5px;

    p {
      font-size: 10px;
      color: white;
      opacity: 0.6;
      text-shadow: 0 0 5px black;
    }
  }
`
const ImageSetWrapper = styled(SectionWrapper)`
  ${ImageSetContainer} {
    width: 70%;
    max-width: 1000px;
    margin: 0 auto;
    color: black;

    img {
      width: 100%;
    }

    ${SlideshowTitle} {
      display: none;
    }
    ${SlideshowCta} {
      flex-direction: column;
      > div {
        padding: 0;
      }
    }
    ${media.md`
      width: 100%;
    `};
  }
`

const StickyHeader = styled.div`
  position: absolute;
  top: 100%;
  width: 100%;
  background: white;
  border-top: ${BORDER_WIDTH}px solid;

  h1 {
    font-size: 40px;
  }
  h2 {
    font-size: 35px;
  }

  ${SectionHeader} {
    padding-bottom: 10px;
    padding-left: 10px;
    padding-right: 10px;
    padding-top: 14px;
  }

  ${media.sm`
    h1 {
      font-size: 30px;;
    }
    h2 {
      font-size: 20px;
    }
  `};
`
