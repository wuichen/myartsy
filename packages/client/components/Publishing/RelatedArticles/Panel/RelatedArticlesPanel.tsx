import { Sans } from "@artsy/palette"
import { track } from "Artsy/Analytics"
import * as Schema from "Artsy/Analytics/Schema"
import { RelatedArticlePanelData } from "Components/Publishing/Typings"
import { once } from "lodash"
import React from "react"
import Waypoint from "react-waypoint"
import styled from "styled-components"
import { RelatedArticlesPanelLink } from "./RelatedArticlesPanelLink"

interface RelatedArticlesPanelProps extends React.HTMLProps<HTMLDivElement> {
  label?: string
  articles: RelatedArticlePanelData[]
}

@track({
  // TODO: reevalutate double naming of context/schema
  context_module: Schema.ContextModule.RelatedArticles,
  subject: Schema.Subject.RelatedArticles,
})
export class RelatedArticlesPanel extends React.Component<
  RelatedArticlesPanelProps
> {
  static defaultProps = {
    label: "Related Stories",
  }

  @track(props => ({ action_type: Schema.ActionType.Impression }))
  trackRelatedImpression() {
    // noop
  }

  render() {
    const { articles, label } = this.props

    return (
      <RelatedArticlesContainer>
        <Label size="3t" weight="medium">
          {label}
        </Label>
        <Waypoint onEnter={once(this.trackRelatedImpression.bind(this))} />
        <Collection>
          {articles.map((article, i) => (
            <RelatedArticlesPanelLink
              article={article}
              key={`relatedArticles-${i}`}
            />
          ))}
        </Collection>
      </RelatedArticlesContainer>
    )
  }
}

const RelatedArticlesContainer = styled.div`
  max-width: 360px;
`

const Collection = styled.div`
  display: flex;
  flex-direction: column;
`

const Label = styled(Sans)`
  margin-bottom: 10px;
`
