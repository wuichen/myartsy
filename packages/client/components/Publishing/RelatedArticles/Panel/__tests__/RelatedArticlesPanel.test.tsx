import { mockTracking } from "Artsy/Analytics"
import { RelatedPanel } from "Components/Publishing/Fixtures/Components"
import { mount } from "enzyme"
import "jest-styled-components"
import React from "react"
import renderer from "react-test-renderer"
import Waypoint from "react-waypoint"
import { RelatedArticlesPanel } from "../RelatedArticlesPanel"
import { RelatedArticlesPanelLink } from "../RelatedArticlesPanelLink"

jest.unmock("react-tracking")

describe("RelatedArticlesPanel", () => {
  const getWrapper = props => {
    return mount(<RelatedArticlesPanel {...props} />)
  }

  let testProps
  beforeEach(() => {
    testProps = { articles: RelatedPanel }
  })

  it("renders the related articles panel", () => {
    const related = renderer
      .create(<RelatedArticlesPanel articles={RelatedPanel} />)
      .toJSON()
    expect(related).toMatchSnapshot()
  })

  it("renders a default label", () => {
    const component = getWrapper(testProps)
    expect(component.text()).toMatch("Related Stories")
  })

  it("renders a provided label", () => {
    testProps.label = "The Best Stories"
    const component = getWrapper(testProps)
    expect(component.text()).toMatch("The Best Stories")
  })

  it("renders links", () => {
    const component = getWrapper(testProps)
    expect(component.find(RelatedArticlesPanelLink)).toHaveLength(3)
  })

  it("tracks link clicks", () => {
    const { Component, dispatch } = mockTracking(RelatedArticlesPanel)
    const component = mount(<Component articles={RelatedPanel} />)
    component
      .find(RelatedArticlesPanelLink)
      .at(0)
      .simulate("click")

    expect(dispatch).toBeCalledWith({
      action_type: "Click",
      context_module: "Related articles",
      subject: "Related articles",
      destination_path:
        "/article/artsy-editorial-15-top-art-schools-united-states",
      type: "thumbnail",
    })
  })

  it("Calls a tracking impression", () => {
    const { Component, dispatch } = mockTracking(RelatedArticlesPanel)
    const component = mount(<Component articles={RelatedPanel} />)
    component
      .find(Waypoint)
      .getElement()
      .props.onEnter()

    expect(dispatch).toBeCalledWith({
      action_type: "Impression",
      context_module: "Related articles",
      subject: "Related articles",
    })
  })
})
