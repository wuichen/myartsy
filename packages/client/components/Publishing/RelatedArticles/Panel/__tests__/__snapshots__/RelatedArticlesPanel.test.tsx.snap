// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelatedArticlesPanel renders the related articles panel 1`] = `
.c2 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
}

.c6 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 18px;
  line-height: 22px;
  color: #000;
}

.c4 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: left;
  -webkit-justify-content: left;
  -ms-flex-pack: left;
  justify-content: left;
  margin-bottom: 20px;
}

.c5 {
  min-width: 80px;
  height: 55px;
  margin-right: 10px;
}

.c0 {
  max-width: 360px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  margin-bottom: 10px;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
    fontFamily={
      Object {
        "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
        "fontWeight": 500,
      }
    }
    fontSize="14px"
  >
    Related Stories
  </div>
  <span
    style={
      Object {
        "fontSize": 0,
      }
    }
  />
  <div
    className="c3"
  >
    <a
      className="c4"
      href="/article/artsy-editorial-15-top-art-schools-united-states"
      onClick={[Function]}
    >
      <img
        className="c5"
        src="https://d7hftxdivxxvm.cloudfront.net?resize_to=fill&src=https%3A%2F%2Fartsy-media-uploads.s3.amazonaws.com%2F4Tq-iYkN8dOpshFoKRXyYw%252Fcustom-Custom_Size___PoetterHall_Exterior%2Bcopy.jpg&width=160&height=110&quality=95"
      />
      <div
        className="c6"
        color="#000"
        fontFamily="'Adobe Garamond W08', 'adobe-garamond-pro', 'AGaramondPro-Regular', 'Times New Roman', Times, serif"
        fontSize="18px"
      >
        The 15 Top Art Schools in the United States
      </div>
    </a>
    <a
      className="c4"
      href="/article/artsy-editorial-four-years-walter-de-marias-death-final-work-complete"
      onClick={[Function]}
    >
      <img
        className="c5"
        src="https://d7hftxdivxxvm.cloudfront.net?resize_to=fill&src=https%3A%2F%2Fartsy-media-uploads.s3.amazonaws.com%2F6IqxBTQCkExip2auQ7ZWCA%252FDEMAR-2011.0006-B.jpg&width=160&height=110&quality=95"
      />
      <div
        className="c6"
        color="#000"
        fontFamily="'Adobe Garamond W08', 'adobe-garamond-pro', 'AGaramondPro-Regular', 'Times New Roman', Times, serif"
        fontSize="18px"
      >
        Four Years after Walter De Mariaâ€™s Death, His Final Work Is Complete
      </div>
    </a>
    <a
      className="c4"
      href="/news/artsy-editorial-french-art-history-in-a-nutshell"
      onClick={[Function]}
    >
      <img
        className="c5"
        src="https://d7hftxdivxxvm.cloudfront.net?resize_to=fill&src=https%3A%2F%2Fartsy-media-uploads.s3.amazonaws.com%2FlEcCm2XbfZ7bPAVgLlM21w%252Flarger-21.jpg&width=160&height=110&quality=95"
      />
      <div
        className="c6"
        color="#000"
        fontFamily="'Adobe Garamond W08', 'adobe-garamond-pro', 'AGaramondPro-Regular', 'Times New Roman', Times, serif"
        fontSize="18px"
      >
        French Art History in a Nutshell
      </div>
    </a>
  </div>
</div>
`;
